storage/innobase/handler/handler0alter.cc

```diff
@@ -4211,103 +4211,179 @@ innobase_dropping_foreign(
 	return(false);
 }
 
-/** Determines if an InnoDB FOREIGN KEY constraint depends on a
-column that is being dropped or modified to NOT NULL.
+/** Check whether the given column name exists in names
+@param n_fields number of column names
+@param name     column name to search
+@param names    column names to be searched from
+@return true if name exists or false if name doesn't exist. */
+static
+bool col_exists(unsigned n_fields, const char *name, const char **names)
+{
+  for (unsigned i = 0; i < n_fields; i++)
+  {
+    if (!strcmp(names[i], name))
+      return true;
+  }
+  return false;
+}
+
+/** Determines if an InnoDB FOREIGN KEY constraint depends on
+the nullability changes of a column.
+Enforce the following rules:
+
+i) Don't allow the child foreign key column from NULL TO NOT NULL when
+     1) Foreign key constraint type is ON UPDATE SET NULL
+     2) Foreign key constraint type is ON DELETE SET NULL
+     3) Foreign key constraint type is UPDATE CASCADE and referenced
+        column allows NULL values
+
+ii) Don't allow the parent referenced column from NOT NULL to NULL when
+foreign key constraint type is UPDATE CASCADE and foreign column
+doesn't allow NULL values
+
 @param user_table InnoDB table as it is before the ALTER operation
-@param col_name Name of the column being altered
 @param drop_fk constraints being dropped
 @param n_drop_fk number of constraints that are being dropped
-@param drop true=drop column, false=set NOT NULL
+@param col_name  modified column name
+@param new_field_flags Modified field flags
 @retval true Not allowed (will call my_error())
 @retval false Allowed
 */
-MY_ATTRIBUTE((pure, nonnull(1,4), warn_unused_result))
 static
-bool
-innobase_check_foreigns_low(
-	const dict_table_t*	user_table,
-	dict_foreign_t**	drop_fk,
-	ulint			n_drop_fk,
-	const char*		col_name,
-	bool			drop)
+bool innobase_check_foreigns_nullability(
+  const dict_table_t *user_table,
+  dict_foreign_t     **drop_fk,
+  ulint              n_drop_fk,
+  const char*        col_name,
+  uint32_t           new_field_flags)
 {
-	dict_foreign_t*	foreign;
-	ut_ad(mutex_own(&dict_sys.mutex));
-
-	/* Check if any FOREIGN KEY constraints are defined on this
-	column. */
-
-	for (dict_foreign_set::const_iterator it = user_table->foreign_set.begin();
-	     it != user_table->foreign_set.end();
-	     ++it) {
+  ut_ad(mutex_own(&dict_sys.mutex));
 
-		foreign = *it;
+  /* Changing from NOT NULL to NULL. So check referenced set */
+  if (!(new_field_flags & NOT_NULL_FLAG))
+    goto ref_field_check;
```

Comment:
We could remove the `goto` by inverting the condition and adding `{`. With 2 spaces per indentation level (instead of 8-space TAB) there should be no need for such tricks.